src/worker/core/GameSim.basketball/PlayByPlayLogger.ts

```diff
@@ -136,49 +130,12 @@ type PlayByPlayEventInputNoScore =
 			clock: number;
 	  }
 	| {
-			type: "fgaAtRim";
-			t: TeamNum;
-			pid: number;
-			clock: number;
-	  }
-	| {
-			type: "fgaLowPost";
-			t: TeamNum;
-			pid: number;
-			clock: number;
-	  }
-	| {
-			type: "fgaMidRange";
-			t: TeamNum;
-			pid: number;
-			clock: number;
-	  }
-	| {
-			type: "fgaTp";
-			t: TeamNum;
-			pid: number;
-			clock: number;
-			desperation: boolean;
-	  }
-	| {
-			type: "fgaTpFake";
-			t: TeamNum;
-			pid: number;
-			clock: number;
-			desperation: boolean;
-	  }
-	| {
-			type: "fgaTipIn";
-			t: TeamNum;
-			pid: number;
-			pidPass: number;
-			clock: number;
-	  }
-	| {
-			type: "fgaPutBack";
+			type: fgaType;
 			t: TeamNum;
 			pid: number;
 			clock: number;
+			desperation?: boolean | undefined; // needed for tp/tpFake. undefined for others
+			pidPass?: number | undefined; // needed for tipin. undefined for others
```

Comment:
Shorter code is nice, but not at the expense of type correctness. It's worth having 3 separate types here rather than combining it all into one. The whole point of TypeScript is that comments like "needed for tp/tpFake. undefined for others" should not be needed, instead TypeScript handles it for you automatically :)